src/LuauExt.cpp

```diff
@@ -558,22 +581,58 @@ struct FindSymbolReferences : public Luau::AstVisitor
         return Luau::Symbol(name) == symbol;
     }
 
+    bool visitWriteExpr(Luau::AstExpr* expr) const
+    {
+        if (auto local = expr->as<Luau::AstExprLocal>(); local && visitLocal(local->local))
+            return true;
+        else if (auto global = expr->as<Luau::AstExprGlobal>(); global && visitGlobal(global->name))
+            return true;
+        return false;
+    }
+
     bool visit(Luau::AstStatLocalFunction* function) override
     {
         if (visitLocal(function->name))
         {
-            result.push_back(function->name->location);
+            addResult(function->name->location, lsp::DocumentHighlightKind::Write);
         }
         return true;
     }
 
+    bool visit(Luau::AstStatFunction* function) override
+    {
+        auto name = function->name;
+        if (visitWriteExpr(name))
+            addResult(name->location, lsp::DocumentHighlightKind::Write);
+        else
+            name->visit(this);
+        function->func->visit(this);
+        return false;
+    }
+
     bool visit(Luau::AstStatLocal* al) override
     {
         for (size_t i = 0; i < al->vars.size; ++i)
         {
-            if (visitLocal(al->vars.data[i]))
+            auto var = al->vars.data[i];
+            auto isSymbol = visitLocal(var);
+            if (symbol.local && i < al->values.size)
             {
-                result.push_back(al->vars.data[i]->location);
+                if (symbol.local->name == var->name && isRequire(al->values.data[i]))
+                {
+                    if (isSymbol)
+                        isModuleImport = true;
+                    else if (symbol.local->location.begin < var->location.begin)
+                    {
+                        // Stop processing this scope because the symbol got shadowed by a module import
```

Comment:
Similarly to the depth check, this exists for two reasons:
1) it allows us to make the type prefix / module import search sound: when the module import gets shadowed, we shouldn't process the following type prefixes that have the same name as the module name, again simulating locals.
2) it allows us to skip a bit of processing in some cases (doesn't matter that much in this case, though).
Together with the depth check, this should make type prefix references completely sound. I thought I might as well do this since it can save us a bit of processing too.